prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,537 @@
+		goto err;
+	}
+
+	fd = open(filename, O_CREAT | O_RDWR | O_EXCL, 0700);
+	if (fd < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "IO Error: %s\n", strerror(errno));
+		ret = -FI_EIO;
+		goto err1;
+	}
+
+	handle->fd = fd;
+	handle->filename = filename;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Successfully opened: %s with handle : %d\n.",
+		   handle->filename, handle->fd);
+
+	unlink(handle->filename);
+
+	return ret;
+
+err1:
+	free(filename);
+err:
+	return ret;
+}
+
+static size_t __page_count(struct gnix_mbox_alloc_handle *handle)
+{
+	size_t total_size = CEILING((handle->mbox_size * handle->mpmmap),
+				     handle->page_size);
+	size_t page_count;
+
+	page_count = total_size / handle->page_size;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Mbox_size: %zu, mpmmap: %zu, page_size: %zu\n",
+		   handle->mbox_size, handle->mpmmap, handle->page_size);
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Total size: %zu, page_count: %zu\n", total_size,
+		   page_count);
+
+	if (page_count <= 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid size requested, truncating to single page.\n");
+		page_count = 1.0;
+	}
+
+	return page_count;
+}
+
+static size_t __mbox_count(struct gnix_mbox_alloc_handle *handle)
+{
+	size_t mbox_count = (__page_count(handle) * handle->page_size) /
+			    handle->mbox_size;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Mbox_count: %zu.\n", mbox_count);
+	return mbox_count;
+}
+
+static int __create_slab(struct gnix_mbox_alloc_handle *handle)
+{
+	struct gnix_slab *slab;
+	gni_return_t status;
+	size_t total_size;
+	int ret;
+
+	slab = calloc(1, sizeof(*slab));
+	if (!slab) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating slab: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "total_size requested for mmap: %zu.\n",
+		   total_size);
+
+	slab->used = calloc(1, sizeof(*(slab->used)));
+	if (!slab->used) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating bitmap: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	slab->base = mmap(0, total_size, (PROT_READ | PROT_WRITE), MAP_SHARED,
+			  handle->fd, handle->last_offset);
+	if (slab->base == MAP_FAILED) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "%s\n", strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err2;
+	}
+
+	ret = alloc_bitmap(slab->used, __mbox_count(handle));
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error allocating bitmap.\n");
+		goto err3;
+	}
+
+	status = GNI_MemRegister(handle->nic_handle->gni_nic_hndl,
+				 (uint64_t) slab->base, total_size,
+				 handle->cq_handle,
+				 GNI_MEM_READWRITE, -1,
+				 &slab->memory_handle);
+	if (status != GNI_RC_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "GNI_MemRegister failed: %s\n",
+			  gni_err_str[status]);
+		ret = gnixu_to_fi_errno(status);
+		goto err4;
+	}
+
+	slab->allocator = handle;
+
+	slist_insert_tail(&slab->list_entry, &handle->slab_list);
+
+	handle->last_offset += total_size;
+
+	return ret;
+
+err4:
```

Comment:
Nit: Could these have meaningful labels? err_failed_reg, or something of the like.